File: euler/94/slow_brute_force.py
from math import sqrt, isqrt


def bs(target, f):
    l = 1
    r = sqrt(int(1e9))
    for _ in range(30):
        mid = (l + r) / 2
        if f(mid) < target:
            l = mid
        else:
            r = mid
    return mid

def A1(n):
    return sqrt(1/16*(n+1)**2*(3*n+1)*(n-1))

def A2(n):
    return sqrt(1/16*(n-1)**2*(3*n-1)*(n+1))

# def check1(area, n):
#     return area**2 == 1/16*(n+1)**2*(3*n+1)*(n-1)
    
# def check2(area, n):
#     return area**2 == 1/16*(n-1)**2*(3*n-1)*(n+1)


def check1(n):
    a2_16 = (n+1)**2*(3*n+1)*(n-1)
    if a2_16 % 16 != 0:
        return False
    a2 = a2_16 // 16
    a = isqrt(a2)
    return a * a == a2
    
def check2(n):
    a2_16 = (n-1)**2*(3*n-1)*(n+1)
    if a2_16 % 16 != 0:
        return False
    a2 = a2_16 // 16
    a = isqrt(a2)
    return a * a == a2

# def check(n):

# ans = 0
# # for area in range(1, isqrt(int(1e9))):
# for area in range(1, int(1e7)):
#     n = round(bs(area, A1))
#     if check1(area, n):
#         print(n)
#         ans += 3*n+1
#     elif check2(area, n):
#         print(n)
#         ans += 3*n-1
# print(ans)

# oh oops this is too slow lol

ans = 0
for n in range(2, int(3e8)):
    if check1(n):
        print(n)
        ans += 3*n+1
    if check2(n):
        print(n)
        ans += 3*n-1
    if n % int(1e7) == 0:
        print(n)
print(ans)

# some solutions:
# 5
# 17
# 65
# 241
# 901
# 3361
# 12545
# 46817
# 174725
# 652081
# 2433601
# 9082321

File: euler/94/slow_brute_force.cpp
// I think I can actually brute force it in reasonable time

#include <algorithm>
#include <bitset>
#include <complex>
#include <deque>
#include <exception>
#include <fstream>
#include <functional>
#include <iomanip>
#include <ios>
#include <iosfwd>
#include <iostream>
#include <istream>
#include <iterator>
#include <limits>
#include <list>
#include <locale>
#include <map>
#include <memory>
#include <new>
#include <numeric>
#include <ostream>
#include <queue>
#include <set>
#include <sstream>
#include <stack>
#include <stdexcept>
#include <streambuf>
#include <string>
#include <typeinfo>
#include <utility>
#include <valarray>
#include <vector>
#include <array>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <forward_list>
#include <future>
#include <initializer_list>
#include <mutex>
#include <random>
#include <ratio>
#include <regex>
#include <scoped_allocator>
#include <system_error>
#include <thread>
#include <tuple>
#include <typeindex>
#include <type_traits>
#include <unordered_map>
#include <unordered_set>

using namespace std;
typedef long long ll;
typedef pair<int, int> pii;
typedef pair<double, double> pdd;
typedef pair<double, int> pdi;
typedef pair<int, double> pid;
typedef tuple<int, int, int> tiii;
typedef vector<int> vi;
typedef vector<ll> vll;
typedef vector<vector<int>> vvi;
typedef vector<vector<ll>> vvll;
typedef vector<double> vd;
#define CONTAINS(set, x) (set.find(x) != set.end())
#define FOR(i, a, b) for(int i=(a);i<(b);i++)
#define ROF(i, a, b) for(int i=(b)-1;i>=(a);--i)
#define fi first
#define se second
#define bk back()
#define ft front()
#define pb push_back
#define all(x) (x).begin(),(x).end()
#define sz(x) x.size()
#define trav(v,x) for(auto& v : x)
const ll MOD = 1e9+7;
template<class T> T gcd(T a,T b){return b?gcd(b,a%b):a;}
template<class T> T modpow(T a,T b,T m){
    T res=1;for(;b;b/=2,a=(a*a)%m)if(b&1)res=(res*a)%m;return res;}
template<class T> T inv(T a,T b){return 1<a?b-inv(b%a,a)*b/a:1;}
void re(int& e){cin>>e;}
void re(ll& e){cin>>e;}
void re(int* v, int n){FOR(i,0,n)cin>>v[i];}
void re(vi& v, int n){FOR(i,0,n)cin>>v[i];}
#define endl ('\n')  // avoid flushing
template <typename T>
ostream& operator<<(ostream& os, const vector<T>& v){
    cout<<"[";FOR(i,0,sz(v))cout<<v[i]<<(i==sz(v)-1?"":", ");cout<<"]";return cout;}
template <typename T, typename U>
ostream& operator<<(ostream& os, const pair<T, U>& v){ cout<<"{"<<v.fi<<", "<<v.se<<"}"; return cout;}

const ll mx = 1e9;

bool check1(ll n) {
    ll a2_16 = (n+1)*(n+1)*(3*n+1)*(n-1);
    if (a2_16 % 16 == 0) {
        ll a2 = a2_16 / 16;
        ll a = sqrt(a2);
        return a * a == a2;
    } else {
        return false;
    }
}

bool check2(ll n) {
    ll a2_16 = (n-1)*(n-1)*(3*n-1)*(n+1);
    if (a2_16 % 16 == 0) {
        ll a2 = a2_16 / 16;
        ll a = sqrt(a2);
        return a * a == a2;
    } else {
        return false;
    }
}

void Solve()
{
    ll ans = 0;
    for (int n = 2; n < 1000; n++) {
        if (check1(n)) {
            cout << n << endl;
            ans += 3*n+1;
        }
        if (check2(n)) {
            cout << n << endl;
            ans += 3*n-1;
        }
        assert(ans < LLONG_MAX / 4);
    }
    cout << ans << endl;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout << fixed << setprecision(6);
    Solve();
    return 0;
}

// it turns out it's still too slow